### PR TITLE
Table Setting + Custom Navbar Addon

### LINKED ISSUE #255: dynamic length options
**Is your feature request related to a problem? Please describe.**

I want to make an addon for the forums where you have predifined strings you can send. I don't want to have to limit the amount to 5 or something and make each string option manually.

**Describe the solution you'd like**

a new option type that had a dynamic length of multi strings if that makes sense.

I'll send a mockup soon™️

**Describe alternatives you've considered**

creating a custom options menu for my addon but that's bad UX, probably not good and technically not exactly very possible

**Additional context**

i send mockup soon

### LINKED ISSUE #721: Wiki Button
**Is your feature request related to a problem? Please describe.**

It's hard to acsess the forums without a navbar button. Same for the Wiki.

**Describe the solution you'd like**

I propose a addon that adds a Wiki button to the navbar, just like Maximouse and World_Language's Discuss button.

### LINKED ISSUE #951: Custom Navbar
**Describe the new addon**

Adds the ability to change the url and text on the Ideas and About buttons 

**IMPORTANT: As W_L requested ANY `javascript:` url will make the addon not run**
The user is warned about this when changing settings

**Link to the new addon**

https://github.com/YOYITsM3M8/ScratchAddons/tree/custom-navbar
(I will be making a pull request if people like this)

### LINKED ISSUE #1333: Removes the 'Tutorials' button
Hello,
when we activate all the functions of the editor, the bar at the top is too big compared to the rest of the editor. I therefore suggest that you make an option that removes the 'Tutorials' button.
thank you,
Miala

Text translated from French by Google Translate

### PR DESCRIPTION
**Pull request by TheColaber**
Resolves #951
Resolves #721
Resolves #255

Exceeds #1342
Related ScratchAddons/manifest-schema#40
Merge after this: ScratchAddons/manifest-schema#50
Merge after this: ScratchAddons/l10n-script#7

### Changes

- turn discuss-button to Custom Navbar
- background logic for when a user goes from using old setting to table settings
- remove background logic for previous versions
- Move handleKeySettings to addon-settings component
- removes forums tag... why do we need it??
- adds images/icons/drag.svg
- adds libraries/thirdparty/sortable.min.js
- simpler way to do checkValidity

### Manifest Spec
```jsonc
{
  "name": String,
  "id": String,
  "type": String,
  "default": Array of Arrays of Strings, Booleans, and Numbers,
  "row": Array of Settings // excluding type table setting without default prop
  "presets": Array of { name: String, values: Arrays of Strings, Booleans, and Numbers }
}
```

### TODO
- [x] add the library info
- [x] description and name
- [x] don't make javascript URLs work
- [x] lightmode
- [x] presets
- [x] i10n - apple said they'd work on this soon

### Tests

https://user-images.githubusercontent.com/72760579/123692567-f7a2dc80-d824-11eb-9ed8-b5a731fa83eb.mp4

### CODE REVIEW

**Comment by mxmou** [comment]
I see this when I open the settings page after reloading the extension. What causes this?
image

Could the delete, add and drag icons have hover backgrounds, just like the reset settings one (but with light mode support)?

Possible name and description:
- Customizable navigation bar
- Choose a list of links to be displayed in the navigation bar.

**Comment by TheColaber** [comment]
What were the settings set to?
```js
JSON.parse(JSON.stringify(vue.addonSettings["discuss-button"]))
```
Any errors in the console?

**Comment by TheColaber** [comment]
That's weird... I didn't create something called `links`... and Items is supposed to be an array, not an object...
image
What are the value of `items`

**Comment by mxmou** [comment]
> What are the value of `items`

image

**Comment by TheColaber** [comment]
I may have messed up the background transition code...

**Comment by TheColaber** [comment]
I was able to repro, and got my settings looking like this:
image
Turns out I messed up my transition code

**Comment by TheColaber** [comment]
@mxmou fixed. you can try setting your settings to `buttonName: Discuss` and `removeIdeasBtn: true` and then reload the extension

**Comment by mxmou** [comment]
> @mxmou fixed. you can try setting your settings to `buttonName: Discuss` and `removeIdeasBtn: true` and then reload the extension

It works now.

Could the background color of items be `#333333` to avoid adding another color that isn't used anywhere on the page yet?

**Comment by TheColaber** [comment]
@mxmou are you fine with the current design of the table? You can commit any changes you want to improve it.

**Comment by TheColaber** [comment]
Also this is weird...

https://user-images.githubusercontent.com/72760579/123882210-4f6c4100-d914-11eb-8a32-5220dcf961f4.mp4

I'm guessing this happens to new addons only? We may need an issue for this.

**Comment by TheColaber** [comment]
mxmou, btw, items in the popup now look like this
image
Is there any way we can try to keep "name" on one line?

**Comment by mxmou** [comment]
> mxmou, btw, items in the popup now look like this
> image
> Is there any way we can try to keep "name" on one line?

It doesn't look like that for me.

**Comment by mxmou** [comment]
The transition code should put Discuss before About.

**Comment by mxmou** [comment]
Reloading the extension seems to make `items` an object.

**Comment by lisa-wolfgang** [comment]
A couple of things I noticed:
- The popup no longer loads... that might be a tiny problem.
- The settings on the settings page also don't load...
- Oh, that's why:
image

**Comment by TheColaber** [comment]
Fixed, I believe

File: addons/discuss-button/addon.json [hunk 1]
```diff
@@ -1,14 +1,6 @@
 {
-  "name": "Discuss button",
-  "description": "Adds the discuss button to the navbar.",
-  "credits": [
```
Comment (RedGuy12) [inline addons/discuss-button/addon.json:4]: was removing credits intentional?

File: addons/discuss-button/userscript.js [hunk 2]
```diff
@@ -1,29 +1,50 @@
...
     }
-  } else {
-    const el = await addon.tab.waitForElement("div#topnav ul.site-nav");
-    if (addon.settings.get("removeIdeasBtn")) el.getElementsByTagName("li")[2].remove();
-    el.insertBefore(link, el.getElementsByTagName("li")[2]);
+    while (!searchItem.previousSibling.classList.contains("logo")) {
+      searchItem.previousSibling.remove();
+    }
+  }
+  async function init() {
```
Comment (RedGuy12) [inline addons/discuss-button/userscript.js:37]: why async?

File: background/get-addon-settings.js [hunk 3]
```diff
@@ -7,49 +7,33 @@ chrome.storage.sync.get(["addonSettings", "addonsEnabled"], ({ addonSettings = {
       const settings = addonSettings[addonId] || {};
       let madeChangesToAddon = false;
       if (manifest.settings) {
-        if (addonId === "editor-dark-mode") {
```
Comment (RedGuy12) [inline background/get-addon-settings.js:10]: maybe we should keep old addon transition code?

File: webpages/settings/style.css [hunk 4]
```diff
@@ -133,6 +133,12 @@ h1 {
   vertical-align: middle;
   filter: invert(0) brightness(10);
   margin-inline-end: 8px;
+  user-drag: none;
```
Comment (RedGuy12) [inline webpages/settings/style.css:136]: image (caniuse)

so remove all instances of `user-drag` and use `draggable="false"` instead

File: webpages/settings/style.css [hunk 5]
```diff
@@ -133,6 +133,12 @@ h1 {
...
+  user-select: none;
+  -moz-user-select: none;
+  -webkit-user-drag: none;
+  -webkit-user-select: none;
+  -ms-user-select: none;
```
Comment (RedGuy12) [inline webpages/settings/style.css:141]: lol not ms

we don't support ie or old edge

File: webpages/settings/style.css [hunk 6]
```diff
@@ -133,6 +133,12 @@ h1 {
   vertical-align: middle;
   filter: invert(0) brightness(10);
   margin-inline-end: 8px;
+  user-drag: none;
+  user-select: none;
+  -moz-user-select: none;
```
Comment (RedGuy12) [inline webpages/settings/style.css:138]: remove prefixed versions, `user-select` covers all browsers we support

**Review by RedGuy12 — CHANGES_REQUESTED** [review]

File: addons/discuss-button/addon.json (hunk 1 above)
Comment (TheColaber) [reply addons/discuss-button/addon.json:4]: yes

File: background/get-addon-settings.js (hunk 3 above)
Comment (TheColaber) [reply background/get-addon-settings.js:10]: why, its from like 5 versions

File: webpages/settings/style.css (hunk 6 above)
Comment (RedGuy12) [reply webpages/settings/style.css:138]: wait, nvm

edge needs -webkit- still
but remove -moz-

File: webpages/settings/style.css (hunk 4 above)
Comment (TheColaber) [reply webpages/settings/style.css:136]: i can remove it but adding `draggable="false"` is uneeded

Comment (RedGuy12) [reply webpages/settings/style.css:136]: it's only supported in chromium with `-webkit-` and no other browsers support it

why is draggable=false unneeded? that's literally what this does

Comment (TheColaber) [reply webpages/settings/style.css]: Because -webkit-user-drag works

Comment (RedGuy12) [reply webpages/settings/style.css]: only in chromium

and it's not reccomended

**Review by RedGuy12 — APPROVED** [review]

File: addons/discuss-button/userscript.js [hunk 7]
```diff
@@ -1,29 +1,50 @@
...
-        markAsSeen: true,
-        reduxCondition: (state) => {
-          if (!state.scratchGui) return true;
-          return state.scratchGui.mode.isPlayerOnly;
-        },
-      });
-      if (addon.settings.get("removeIdeasBtn")) el.getElementsByTagName("li")[3].remove();
-      el.insertBefore(link, el.getElementsByTagName("li")[3]);
+    const a = document.createElement("a");
+    a.href = url.startsWith("javascript") ? "" : url;
```
Comment (apple502j) [inline addons/discuss-button/userscript.js:17]:
```suggestion
    a.href = /^https?:\/\//g.test(url) ? url : "";
```

**Review by apple502j — CHANGES_REQUESTED** [review]

File: addons/discuss-button/userscript.js (hunk 7 above)
Comment (TheColaber) [reply addons/discuss-button/userscript.js:17]: What if I wanted to put the sa settings page url? Why not just exclude js urls?

Comment (apple502j) [reply addons/discuss-button/userscript.js:17]:
```suggestion
    a.href = /^(?:https?|(?:chrome|moz)-extension):\/\//g.test(url) ? url : "";
```

Comment (WorldLanguages) [reply addons/discuss-button/userscript.js:17]: > What if I wanted to put the sa settings page url? Why not just exclude js urls?

I don't think scratch.mit.edu can directly link to chrome-extension://*

**Comment by FunctionalMetatable** [comment]
Pathnames in urls do not work

File: addons/discuss-button/userscript.js (hunk 7 above)
Comment (TheColaber) [reply addons/discuss-button/userscript.js:17]: @apple502j Your commit has broken pathnames.

Comment (TheColaber) [reply addons/discuss-button/userscript.js:17]: Does this work ok?
```suggestion
    a.href = /\s*javascript:/g.test(url) ? "" : url;
```

Comment (RedGuy12) [reply addons/discuss-button/userscript.js:17]: FYI Apple, on edge it's just `extention://` without the prefix

But why exclude js urls anyway?

Comment (TheColaber) [reply addons/discuss-button/userscript.js:17]: Yeah keeping it would be cool. Tho it's a bad idea for many reasons.

Comment (apple502j) [inline addons/discuss-button/userscript.js:17]:
```suggestion
    a.href = ["http:", "https:"].includes(new URL(url, location.href).protocol) ? url : "";
```

**Comment by RedGuy12** [comment]
> Resolves #762

it does not?

**Comment by TheColaber** [comment]
It does, read my comment in that issue.

**Comment by Samq64** [comment]
custom-navbar-addon
I think there should be 2 items per row on bigger windows.

**Comment by RedGuy12** [comment]
Wouldn't it just be display inline-block?


________________________________

**Comment by Samq64** [comment]
I noticed that resetting a value in a table causes it to go blank instead of using the default value even if there is one. Is this intentional?

**Comment by TheColaber** [comment]
@WorldLanguages @mxmou How do you think the ui for presets should be? Maybe something like this?

https://user-images.githubusercontent.com/72760579/143478180-a595ec8b-a1e8-49cb-9031-68afcd3fa312.mp4

**Comment by mxmou** [comment]
> @WorldLanguages @mxmou How do you think the ui for presets should be? Maybe something like this?

There is no indication that clicking one of the buttons adds a new row - I think a dropdown menu next to the add row button would be better.

> Also, what should our presets be?

Discuss and wiki are commonly requested, so those should certainly be there. You could also add the default buttons as presets to allow them to be easily re-added after removing one of them.

**Comment by TheColaber** [comment]
@mxmou Is this similar to what you meant?

https://user-images.githubusercontent.com/72760579/143526341-b38717e5-7cf9-426f-a703-10ec3c2c41e2.mp4

**Comment by mxmou** [comment]
> @mxmou Is this similar to what you meant?

Yes, this is exactly what I meant. I changed the design a bit so that it looks more similar to other buttons.

> Should the inputs be aligned? image

I don't think so. Other settings aren't aligned in the popup either.

> These preset dropdowns are here... They probably shouldn't since the default is "", and future addons probably won't want a preset dropdown either? image

Perhaps the dropdowns could contain the row presets? Future addons might also want to set the default values for a new row, and the ability to reset to those values would be useful.

**Comment by TheColaber** [comment]
> Perhaps the dropdowns could contain the row presets? Future addons might also want to set the default values for a new row, and the ability to reset to those values would be useful.

Hmm... It would be better if it reset the entire row, and not just one input in the row? Is there a good way to do that?

**Comment by CST1229** [comment]
Are tables of colors are supported? It would be useful for stuff like a color swatches addon where you can customize the colors.

**Comment by TheColaber** [comment]
> Are tables of colors are supported? It would be useful for stuff like a color swatches addon where you can customize the colors.

All setting types should be supported
image

**Review by apple502j — APPROVED** [review]

**Comment by mxmou** [comment]
> > Perhaps the dropdowns could contain the row presets? Future addons might also want to set the default values for a new row, and the ability to reset to those values would be useful.
> 
> Hmm... It would be better if it reset the entire row, and not just one input in the row? Is there a good way to do that?

If the reset button is separate for each input, it isn't very difficult to reset an entire row - just click the reset button for every input.

**Review by mxmou — APPROVED** [review]

**Comment by TheColaber** [comment]
It's kind of hard to add the dropdowns since the way presets work for tables and all settings are completely different...

File: addons/discuss-button/addon.json [hunk 8]
```diff
@@ -1,14 +1,6 @@
...
-    {
-      "name": "Maximouse"
-    },
-    {
-      "name": "World_Languages"
-    }
-  ],
+  "name": "Customizable navigation bar",
+  "description": "Choose a list of links to be displayed in the navigation bar.",
```
Comment (RedGuy12) [inline addons/discuss-button/addon.json:3]: Sounds off

File: libraries/license-info.json [hunk 9]
```diff
@@ -10,5 +10,6 @@
   "color-picker-web-component": "MIT",
   "comlink": "Apache-2.0",
   "Sora": "OFL",
+  "sortable": "MIT",
```
Comment (RedGuy12) [inline libraries/license-info.json:13]: Don't forget to update the LICENSES.md

**Comment by TheColaber** [comment]
@mxmou It seems that addon-buttons are causing the popup scrollbar to hide a bit...
image
It seems to be caused by https://github.com/ScratchAddons/ScratchAddons/pull/2875/files#diff-55c7d01e0e4e387212119cc7d7e2a0e8137c5a6f0ea7ce8b4ac9a87db76dea77R181 but I'm not sure the proper fix.

**Comment by mxmou** [comment]
> @mxmou It seems that addon-buttons are causing the popup scrollbar to hide a bit...

Fixed. It wasn't actually caused by `addon-buttons`.

**Comment by Samq64** [comment]
Will this make it for 1.23?

**Comment by lisa-wolfgang** [comment]
I saw the commit and my first reaction was "he didn't..." and then I realized that you actually didn't lol

File: addons/discuss-button/addon.json [hunk 10]
```diff
@@ -18,21 +10,58 @@
...
+          "name": "Wiki",
+          "values": { "name": "Wiki", "url": "https://scratch-wiki.info/" }
+        }
+      ]
     }
   ],
-  "versionAdded": "1.0.0",
-  "tags": ["community", "forums"],
+  "versionAdded": "1.23.0",
+  "tags": ["community", "recommended"],
```
Comment (WorldLanguages) [inline addons/discuss-button/addon.json:65]: Featured, not recommended

versionAdded should stays as 1.0.0 or maybe we can discuss whether 1.24.0 fits

**Comment by Mbrick2** [comment]
IT HAS BEEN MERGED

**Comment by Secret-chest** [comment]
WIKI LINK SOON!

**Comment by Secret-chest** [comment]
NO MORE FORCED ABOUT PAGE LINK!

**Pull request merged**